### DISCUSSION TITLE
Run from NEXT LINE

### DISCUSSION THREAD

**Discussion by Gagit41** [vlachoudis/bCNC#1837] [Q&A]
There appear to be several previous request for way to start G-Code from a line other than the beginning, but I have not seen any suggested solutions. Has this been implement and I just don't know how to use it?  Having to run the code up to the failure point to enable a restart is beyond a pain and erasing the used code can also be painful.  bCNC makes it easy to find the line you want to start at, what is so HARD about creating a resume/restart button for that point in the code???

**Comment by Harvie** [comment]
Nobody implemented it so far.
For some use cases it might be enough to use checkboxes to disable g-code blocks that were already cut.

If i remember correctly it's bit tricky to restart from given line, because grbl nor g-code is completely stateless. eg.: each G1 line tells you where to move, but not from where to move, so you will need to do some g-code processing before you can implement such feature. also there are some offsets, relative coordinates and features like that in g-code, which need to be handled in order for such feature to be reliable.

Therefore nobody yet tackled this. Might be interresting to take a look on how other systems solve this issue.

In the meantime working with g-code blocks is the next best thing, since each g-code block is mostly self-contained with proper setup.

**Reply by Gagit41** [reply]
I think it is pretty well understood that running g-code from the next line can lead to some very undesirable results, but for MANY use cases it can be very successfully.  Many 2D projects are relatively easy to backtrack a few lines and find a good starting place.  In my use case (DIY quilting machine) there is NO z-axis to even worry about.  You would just need to go into the editor, click the spot in the design to find the appropriate place in the g-code and have a way to run code from that line.  
I use LinuxCNC on my CNC router and you can highlight a line of code a choose "Run from next line."  I don't use Mach 3 or 4 but I believe Mach 3 has that sort of feature as well.  
IMO going real basic with the implementation like LinuxCNC has done would please a great many users.

**Comment by Gagit41** [comment]
I think it is pretty well understood that running g-code from the next 
line can lead to some very undesirable results, but for MANY use cases 
it can be very successful. Many 2D projects are relatively easy to 
backtrack a few lines and find a good starting place. In my use case 
(DIY quilting machine) there is NO z-axis to even worry about. You would 
just need to go into the editor, click the spot in the design to find 
the appropriate place in the g-code and have a way to run code from that 
line.
I use LinuxCNC on my CNC router and you can highlight a line of code a 
choose "Run from next line." I don't use Mach 3 or 4 but I believe Mach 
3 has that sort of feature as well.
IMO going real basic with the implementation like LinuxCNC has done 
would please a great many users.

**Comment by Harvie** [comment]
Well. if you submit reasonable PR, i will surely give it a chance.
I did once tried to implement this myself and then it seemed like not being worth the hassle, or got sidetracked by something else.

I tried to modify the sender to skip N lines of g-code, but then the problem was that lines in g-code file and lines in sender do not match, since bCNC does preprocessing of g-code for various reasons and things got messy quite fast.

**Reply by Gagit41** [reply]
I'm afraid I am not even close to being able to code a modification like this to submit on a PR.  But, maybe if the sender could be flushed and then just send the remainder of the g-code file it would resume as intended (position history etc LOST).
Thanks for you timely reply.

**Comment by rar8000** [comment]
Hi. I submitted a PR #1841 that can do this. I tested it with a laser machine. Using the PR branch. You can select the block or line you want to start from and type RES or RESTART. Then run. It will inject 2 gcode statements for feedrate and the XYZ coordinates. This is possible due to a new data processing step and cache. This feature is great for 2D work like mentioned above. 

Also in this PR is the ability to visualize when the spindle is above a threshold S value and/or below a Z value.
